[Fix] 태그 예외처리 추가

dudxor4587 opened the pull request:
# 설명
- 공통 예외처리를 적용하여 태그 예외처리 추가

# 작업내용
- TagRequest 파일에 유효성 처리 추가
- TagNotFoundException 클래스를 추가하여 예외처리
- TagService 클래스에서 get 요청시 ID 존재 여부에 따른 예외처리 적용

dudxor4587 committed:
[fix] 태그 예외처리 추가

boyunee reviewed (approved):
수고하셨습니다!

dudxor4587 committed:
[fix] Transactional 애너테이션 추가 및 주석 추가

hyeond0 commented on src/main/java/shop/bookbom/shop/domain/tag/dto/TagRequest.java:
```diff
@@ -8,4 +8,16 @@
+
+    public TagRequest(String name, Status status) {
+        //이름과 상태의 유효성 검증
+        if (name == null) {
```
제 생각에는 컨트롤러에서 `@RequestBody`로 요청할 때는 이 유효성 검증 로직이 작동하지 않을 것 같은데 
혹시 결과가 어떻게 나오실까요?
또 `@Valid`나 `@Validated` 어노테이션을 사용하지 않으신 이유도 궁금합니다!

dudxor4587 replied:
유효성 검증 로직은 잘 작동했고, @Valid 사용해서 수정사항 반영했습니다!

hyeond0 commented on src/main/java/shop/bookbom/shop/domain/tag/service/TagService.java:
```diff
@@ -12,6 +15,7 @@ public class TagService {
+    @Transactional
     public void saveTagService(String name, Status status) {
         //받아온 새로운 Tag 엔티티 생성
         Tag tag = new Tag(name, status);
```
저희 컨벤션이 엔티티를 생성할 때는 Builder 패턴을 사용하기로 했어서 통일해주셨으면 좋겠습니다 :)

dudxor4587 replied:
수정사항 반영했습니다!

hyeond0 reviewed (approved):
고생하셨습니다!
몇 가지 적어봤어요! 한번 보시고 답변주시면 감사하겠습니다 :)

minsu11 commented on src/main/java/shop/bookbom/shop/domain/tag/service/TagService.java:
```diff
@@ -22,6 +24,10 @@ public void saveTagService(String name, Status status) {
     //태그 삭제
     public void deleteTagService(long tagId) {
         //레포지토리에 접근해 삭제 실행
+        Optional<Tag> tag = tagRepository.findById(tagId);
+        if (tag.isEmpty()) {
+            throw new TagNotFoundException();
+        }
```
Optional 메서드 중 존재하지 않으면 예외처리를 해줄 수 있는 메서드가 있습니다! Optional에 대해서 공부해보시고 한 번 고려해보시면 좋을 것 같습니다!

dudxor4587 replied:
orElseThrow를 이용해서 다시 작성하고 반영했습니다!

minsu11 reviewed (commented):
수고하셨습니다!

minsu11 commented on src/main/java/shop/bookbom/shop/domain/tag/dto/TagRequest.java:
```diff
@@ -8,4 +8,15 @@
 public class TagRequest {
     String name;
```
private를 뺀 이유가 있으실까용?

dudxor4587 replied:
private가 있으면 never assigned 가 발생해서 제거했는데, Data 애너테이션 추가 후 private를 추가하여 반영했습니다!

dudxor4587 committed:
[fix] 수정사항 반영

설명:
- valid 어노테이션 적용
- Builder 패턴 적용
- Optional 예외처리 적용
- private 추가

작업:
- TagRequest에 유효성 검증 애너테이션 추가
- Service에 Builder패턴으로 Tag 생성
- Optional 예외처리 적용
- TagRequest에 private 추가

minsu11 commented on src/main/java/shop/bookbom/shop/domain/tag/dto/TagRequest.java:
```diff
@@ -1,11 +1,19 @@
 
 //태그 등록을 요청받았을 때 넘어온 정보를 받아오는 dto
 @Getter
+@Data
```
@Data가 필요한 어노테이션인가요?? @Data가 어떻게 구현이 되어 있는지 공부해보시면 좋을 것 같습니다!

dudxor4587 replied:
DaTa애너테이션의 대부분의 기능이 사용되지 않아서 Getter만 두고 다시 반영했습니다!

minsu11 commented on src/main/java/shop/bookbom/shop/domain/tag/dto/TagRequest.java:
```diff
@@ -1,11 +1,19 @@
-    String name;
-    Status status;
+    // 유효성 검증
+    @NotNull
```
빈 값과 공백이 잡히나요?? 또한 엄청 긴 문자열이 들어오면 어떻게 되나요??

dudxor4587 replied:
공백은 잡히지 않아서 공백이 잡힐 수 있도록 @NotBlank를 추가하고, 글자 사이 공백이 들어오면 공백을 제거하는 코드를 추가했습니다! 현재는 긴 문자열이 들어오게 되면 그대로 반영됩니다.

minsu11 commented on src/main/java/shop/bookbom/shop/domain/tag/dto/TagRequest.java:
```diff
@@ -1,11 +1,19 @@
+    // 유효성 검증
+    @NotNull
+    private String name;
+    @Enumerated(EnumType.STRING)
```
프로젝트 5주차에 보면 attributeConvertor에 대해서 공부해보면 좋을 것 같습니다!

dudxor4587 replied:
attributeConvertor에 대해서 공부해보았는데, 이 경우에서 어떻게 사용을 해야하는지 잘 모르겠습니다ㅜ

minsu11 commented on src/main/java/shop/bookbom/shop/domain/tag/service/TagService.java:
```diff
@@ -12,16 +15,22 @@ public class TagService {
 
     //태그 등록
+    @Transactional
     public void saveTagService(String name, Status status) {
```
기존에 태그가 있는 경우에 어떻게 되는 건가요??

dudxor4587 replied:
기존 태그가 존재할 경우에 예외를 발생시키는 코드를 추가하였습니다!

minsu11 reviewed (commented):
수고하셨습니다!

dudxor4587 committed:
[fix] 2차 수정사항 반영
설명:
- DaTa 애너테이션 제거
- 유효성 검증 추가
- 존재하는 태그 추가 시 오류 발생
- develop에서 pull받은 내용 추가

작업:
- Data 애너테이션을 제거하고 Getter만 남겨두었습니다.
- 유효성 검증시에 @NotBlank를 추가하고, 사이 공백이 입력되었을 때 공백을 제거하는 코드도 추가하였습니다.
- 이미 존재하는 태그일 경우 TagAlreadyExistException을 발생하게 코드를 작성하였습니다.
